README.md
# node-bitbucket-v2
node.js library to access the Bitbucket API v2

## usage
Not supported for Node < 7.6

```
const Bitbucket = require('node-bitbucket-v2');
const bitbucketApi = new Bitbucket(options);
bitbucketApi.authenticateOAuth2(accessTokenString);

bitbucketApi.user.get().then(({ body }) => {
  console.log(body.uuid);
});
```

### `options`
It is not necessary to provide any options at all (`Bitbucket` can be constructed with no argument).
 - `requesterFn` (`(options) => Promise<any>`): If provided, requests will be made using the function you provide. This is allows you to use your preferred http client. The `options` provided are `{ headers, hostname, method, path, query, url, body? }`. `body` is only provided on `POST` methods. In the case of `getNextPage`, `getPreviousPage`, `getForksFromResponse` and `getParentFromResponse`, only `{ headers, method, url }` are provided in the options. Example:
 ```
  const axios = require('axios');
  const Bitbucket = require('node-bitbucket-v2');

  const requesterFn = (options) => {
    const { url, method, body } = options;

    if (method === 'POST') {
      return axios.post(url, body);
    }

    return axios.get(url);
  };

  const bitbucketApi = new Bitbucket({ requesterFn });
 ```
 - `proxy` (`String`): Defines a proxy to make requests against, instead of `api.bitbucket.org:443`. This option is _ignored_ when `requesterFn` is provided.

For implemented methods, check `bitbucket/repositories.js` and `bitbucket/user.js`.


CHANGELOG.md
# CHANGLOG

## 0.6.0
Thanks to [@andrewyalung](https://github.com/andrewyalung/) for authoring most of this version!
 - `requesterFn` has been added as an option. When `requesterFn` is provided, all request will be processed by that function.
 - `useXhr` has been removed and replaced with `requesterFn`.
 - Fixed non-default `port` being ignored in default request implementation.

## 0.5.2
 - Functions that take an API response as an argument now properly handle taking the _entire_ response, as they did in `0.4.x`. They _also_ still accept just the response's `body` property, as in `0.5.1`. Affected functions: `hasNextPage`, `hasPreviousPage`, `getNextPage`, `getPreviousPage`, `repositories.getForksFromResponse`, `repositories.hasParent`, `repositories.getParentFromResponse`.

## 0.5.1
This version is targeted at two main goals:
1. Bitbucket's _breaking_ API changes (migrating from users and teams to "workspaces"), in order to comply with GDPR. For more details, see: https://developer.atlassian.com/cloud/bitbucket/bitbucket-api-changes-workspaces/
2. Modernizing the code from node-callbacks to promises.

Due to the aforementioned breaking changes on Bitbucket's end, usage of previous versions of this package will stop working in general after 2020-04-29, except for a few basic uses like getting the authenticated user.

 - Started CHANGELOG document.
 - Added Node 7.6 requirement.
 - All asynchronous functions now return a promise, and no longer accept a callback.
 - On success, the old response will now be packed inside a `body` property, and it now has a sibling `statusCode` property.
 - On errors that result from API responses, the response body will now be on a property called `body` (instead of `msg`), and `status` has been renamed to `statusCode`.
 - When `useXhr` is enabled, the entire XHR response will now be provided on both success and failure. Note that this is consistent with the `body` and `statusCode` properties previously described.
 - `repositories.getByUser` and `repositories.getByTeam` have been merged to `repositories.getByWorkspace`.
 - `teams.get` has been removed.
 - `workspaces.get` has been added.
 - Fixed a bug where double JSON parsing would cause parsing to throw an error. Thanks to @mateusmcg for the fix!


bitbucket/index.js
const constants = require('./constants');
const buildRepositories = require('./repositories');
const buildRequest = require('./request');
const buildUser = require('./user');
const buildWorkspaces = require('./workspaces');
const { extractResponseBody } = require('./helpers');


/**
 * Simple JavaScript Bitbucket API v2
 *
 * Based on the PHP GitHub API project http://github.com/ornicar/php-github-api
 */

module.exports = function Bitbucket({ proxy, requesterFn } = {}) {
  /**
   * Define HTTP proxy in format localhost:3128
   */
  let $proxy_host;
  let $proxy_port;
  if (proxy) {
    $proxy_host = proxy.split(':')[0];
    $proxy_port = proxy.split(':')[1];
  }

  const apiModel = {
    $proxy_host,
    $proxy_port,
    constants
  };

  apiModel.repositories = buildRepositories(apiModel);
  apiModel.request = buildRequest({
    proxy_host: $proxy_host,
    proxy_port: $proxy_port,
    requester_fn: requesterFn
  });
  apiModel.user = buildUser(apiModel);
  apiModel.workspaces = buildWorkspaces(apiModel);

  /**
   * Authenticate a user for all next requests using an API token
   *
   * @param {String} accessToken
   * @return {BitbucketApi}        fluent interface
   */
  apiModel.authenticateOAuth2 = (accessToken) => {
    apiModel.request
      .setOption('login_type', 'oauth2')
      .setOption('oauth_access_token', accessToken);

    return apiModel;
  };

  /**
   * Deauthenticate a user for all next requests
   *
   * @return {BitbucketApi}               fluent interface
   */
  apiModel.deAuthenticate = () => {
    apiModel.request
        .setOption('login_type', 'none');

    return apiModel;
  };

  /**
   * Call any route, GET method
   * Ex: api.get('repos/show/my-username/my-repo')
   *
   * @param {String}  route            the Bitbucket route
   * @param {Object}  parameters       GET parameters
   * @param {Object}  requestOptions   reconfigure the request
   */
  apiModel.get = (route, parameters, requestOptions) =>
    apiModel.request.get(route, parameters || {}, requestOptions);

  /**
   * Call any route, DELETE method
   * Ex: api.delete('repos/show/my-username/my-repo')
   *
   * @param {String}  route            the Bitbucket route
   * @param {Object}  parameters       GET parameters
   * @param {Object}  requestOptions   reconfigure the request
   */
  apiModel.delete = (route, parameters, requestOptions) =>
    apiModel.request.delete(route, parameters, requestOptions);

  /**
   * Call any route, POST method
   * Ex: api.post('repos/show/my-username', {'email': 'my-new-email@provider.org'})
   *
   * @param {String}  route            the Bitbucket route
   * @param {Object}  parameters       POST parameters
   * @param {Object}  requestOptions   reconfigure the request
   */
  apiModel.post = (route, parameters, requestOptions) =>
    apiModel.request.post(route, parameters || {}, requestOptions);

  /**
   * Check for whether we can iterate to another page using this.getNextPage(response).
   * @param {response} response A response that was received from the API, or its `body` property.
   * @return {boolean} true if the response indicates more pages are available, false otherwise.
   */
  apiModel.hasNextPage = (response) => Boolean(extractResponseBody(response).next);

  /**
   * Check for whether we can iterate to another page using this.getPreviousPage(response).
   * @param {response} response A response that was received from the API, or its `body` property.
   * @return {boolean} true if the response indicates a previous pages is available, false otherwise.
   */
  apiModel.hasPreviousPage = (response) => Boolean(extractResponseBody(response).previous);

  /**
   * Takes a response and makes an API request for the response's next page.
   * NOTE this should only be called guarded behind a check to `this.hasNextPage(response)`!
   *
   * @param {response} response A response that was received from the API, or its `body` property.
   */
  apiModel.getNextPage = (response) => {
    if (!apiModel.hasNextPage(response)) {
      throw new Error(
        'getNextPage: argument has no next page url. Call hasNextPage first to guard this method call.'
      );
    }

    return apiModel.request.doPrebuiltSend(extractResponseBody(response).next);
  };

  /**
   * Takes a response and makes an API request for the response's previous page.
   * NOTE this should only be called guarded behind a check to `this.hasPreviousPage(response)`!
   *
   * @param {response} response A response that was received from the API, or its `body` property.
   */
  apiModel.getPreviousPage = (response) => {
    if (!apiModel.hasPreviousPage(response)) {
      throw new Error(
        'getPreviousPage: argument has no next page url. Call hasPreviousPage first to guard this method call.'
      );
    }

    return apiModel.request.doPrebuiltSend(extractResponseBody(response).previous);
  };

  return apiModel;
};


bitbucket/repositories.js
const _ = require('lodash');

const constants = require('./constants');
const { extractResponseBody } = require('./helpers');

/**
 * API docs: https://confluence.atlassian.com/bitbucket/repositories-endpoint-423626330.html
 *           https://confluence.atlassian.com/bitbucket/repository-resource-423626331.html
 */
module.exports = function RepositoriesApi(api) {
  return {
    /**
     * Create a new repository
     * @param {String} workspace workspace UUID or slug
     * @param {Object} repo repo metadata as specified by Bitbucket's API documentation.
     *                         NOTE Unlike the normal API, Including an explicit name property in repo is REQUIRED!!
     *                         Due to limitations in the API, the slug is derived from the repo name within this method.
     */
    create: (workspace, repo) => {
      if (!repo || !_.isBoolean(repo.is_private) || !_.isString(repo.name)) {
        throw new Error('Repo must be initialized with a booelan privacy setting and a string name');
      }

      // NOTE the below comment has been clarified by TJ Kells at bitbucket - slugification is implemented here:
      // https://docs.djangoproject.com/en/1.11/_modules/django/utils/text/#slugify
      // Someday the below could be updated to precisely re-implement django's slugification.

      // The official API error is that slugs must be alphanumeric with underscore, dot, and dash, lowercase, and
      // no whitespace. Most things convert to dashes with Atlassian's secret converter but apostophes just disappear
      // (here I've assumed quotes are the same).
      // There are additional constraints not provided in the error message nor documented anywhere that can only be
      // found by trial and error. Among these are: no consecutive dashes except in some weird trivial edge cases
      // (i.e. all dashes, which we won't worry about), no ending in a dash, and very likely no starting in a dash.
      const repoSlug = repo.name
        .replace(/['"]/g, '')
        .replace(/\W/g, '-')
        .replace(/--+/g, '-')
        .replace(/^-/, '')
        .replace(/-$/, '')
        .toLowerCase();

      return api.post(
        `repositories/${encodeURI(workspace)}/${encodeURI(repoSlug)}`,
        repo
      );
    },

    /**
     * Create a new pull request
     *
     * @param {String} workspace workspace UUID or slug
     * @param {String} repoSlug (name) of the repo.
     * @param {Object} pullRequest The PR POST body as specified by Bitbucket's API documentation
     */
    createPullRequest: (workspace, repoSlug, pullRequest) => api.post(
      `repositories/${encodeURI(workspace)}/${encodeURI(repoSlug)}/pullrequests`,
      pullRequest
    ),

    /**
     * Get the info for a single repo
     *
     * @param {String} workspace workspace UUID or slug
     * @param {String} slug (name) of the repo.
     */
    get: (workspace, repoSlug) => api.get(`repositories/${encodeURI(workspace)}/${encodeURI(repoSlug)}`),

    /**
     * Get the branch info for a single repo
     *
     * @param {String} workspace workspace UUID or slug
     * @param {String} slug (name) of the repo.
     */
    getBranches: (workspace, repoSlug) =>
      api.get(`repositories/${encodeURI(workspace)}/${encodeURI(repoSlug)}/refs/branches`),

    /**
     * Get a single commit
     * @param {String} workspace workspace UUID or slug
     * @param {String} slug (name) of the repo.
     * @param {String} the sha of the commit
     */
    getCommit: (workspace, repoSlug, sha) =>
      api.get(`repositories/${encodeURI(workspace)}/${encodeURI(repoSlug)}/commit/${sha}`),

    /**
     * Get the pull requests for a single repo
     *
     * @param {String} workspace workspace UUID or slug
     * @param {String} slug (name) of the repo.
     * @param {constants.pullRequest.states or Array thereof} The PR state. If invalid or undefined, defaults to OPEN
     */
    getPullRequests: (workspace, repoSlug, state) => {
      let stateArray = state;
      if (!stateArray) {
        stateArray = [constants.pullRequest.states.OPEN];
      }
      else if (!_.isArray(stateArray)) {
        stateArray = [stateArray];
      }

      const hasInvalidState = _.find(state, (stateElement) => !_.includes(constants.pullRequest.states, stateElement));
      if (hasInvalidState) {
        stateArray = [constants.pullRequest.states.OPEN];
      }

      const apiParameters = {
        state: stateArray.join(',')
      };

      return api.get(
        `repositories/${encodeURI(workspace)}/${encodeURI(repoSlug)}/pullrequests`,
        apiParameters
      );
    },

    /**
     * Get the pull requests for a single repo, with the destination and source repos on each pull requests totally
     * populated.
     *
     * @param {String} workspace workspace UUID or slug
     * @param {String} repoSlug (name) of the repo.
     * @param {Object} options The fields to populate, and optionally the PR state (defaults to OPEN)
     */
    getPullRequestsWithFields: (workspace, repoSlug, { state, fields } = {}) => {
      if (!_.isArray(fields) || fields.length < 1) {
        throw new Error('getPullRequestsWithFields: options argument missing or has missing/empty \'fields\' array.');
      }

      let stateArray = state;
      if (!stateArray) {
        stateArray = [constants.pullRequest.states.OPEN];
      }
      else if (!_.isArray(stateArray)) {
        stateArray = [stateArray];
      }

      const hasInvalidState = _.find(state, (stateElement) => !_.includes(constants.pullRequest.states, stateElement));
      if (hasInvalidState) {
        stateArray = [constants.pullRequest.states.OPEN];
      }

      const apiParameters = {
        state: stateArray.join(',')
      };

      const fieldsWithEncodedPlus = fields.map((field) => `+${field}`);
      apiParameters.fields = fieldsWithEncodedPlus.join(',');

      return api.get(
        `repositories/${encodeURI(workspace)}/${encodeURI(repoSlug)}/pullrequests`, // eslint-disable-line max-len
        apiParameters
      );
    },

    /**
     * Get the repositories of a workspace
     *
     * @param {String} workspace workspace UUID or slug
     */
    getByWorkspace: (workspace) => api.get(`repositories/${encodeURI(workspace)}`),

    /**
     * Get the forks for a repo
     *
     * @param {String} workspace workspace UUID or slug
     * @param {String} repoSlug (name) of the repo.
     */
    getForks: (workspace, repoSlug) => api.get(`repositories/${encodeURI(workspace)}/${encodeURI(repoSlug)}/forks`),

    /**
     * Get the forks for a repo using an API response that has repository links
     *
     * @param {Object} response API response, or its `body` property
     */
    getForksFromResponse: (response) => {
      const prebuiltURL = _.get(extractResponseBody(response), ['links', 'forks', 'href']);

      if (!prebuiltURL) {
        throw new Error('getForksFromResponse: argument has no \'forks\' url.');
      }

      return api.request.doPrebuiltSend(prebuiltURL);
    },

    /**
     * Get the parent for a repo using an API response that has repository links.
     * This should only be called after a check to hasParent().
     *
     * @param {Object} response API response, or its `body` property
     */
    getParentFromResponse: (response) => {
      const prebuiltURL = _.get(extractResponseBody(response), ['parent', 'links', 'self', 'href']);

      if (!prebuiltURL) {
        throw new Error(
          'getForksFromResponse: argument has no \'parent\' info. Call hasParent first to guard this method call.'
        );
      }

      return api.request.doPrebuiltSend(prebuiltURL);
    },

    /**
     * Determines whether or not the given response has an accessible parent.
     *
     * @param {Object} response API response, or its `body` property
     * @return {boolean} true if the argument has an associated "parent" (i.e. the response is a fork), false otherwise.
     */
    hasParent: (response) => Boolean(extractResponseBody(response).parent)
  };
};


bitbucket/request.js
const _ = require('lodash');
const https = require('https');
const querystring = require('querystring');
const url = require('url');

/**
 * Performs requests on Bitbucket API.
 */
module.exports = function Request(_options) {
  const $defaults = {
    protocol: 'https',
    path: '/2.0',
    hostname: 'api.bitbucket.org',
    format: 'json',
    user_agent: 'node-bitbucket-v2 (https://www.npmjs.com/package/bitbucket-v2)',
    http_port: 443,
    timeout: 20,
    login_type: 'none',
    username: null,
    password: null,
    api_token: null,
    oauth_access_token: null,
    proxy_host: null,
    proxy_port: null,
    requester_fn: null
  };
  const $options = _.defaults({}, _options, $defaults);

  const result = {
    $defaults,
    $options
  };

  return _.assign(result, {
    /**
     * Change an option value.
     *
     * @param {String} name   The option name
     * @param {Object} value  The value
     *
     * @return {Request} The current object instance
     */
    setOption(name, value) {
      $options[name] = value;
      return result;
    },

    /**
      * Get an option value.
      *
      * @param  string $name The option name
      *
      * @return mixed  The option value
      */
    getOption(name, _defaultValue) {
      const defaultValue = _defaultValue === undefined ? null : _defaultValue;
      return $options[name] ? $options[name] : defaultValue;
    },

    /**
     * Send a GET request
     * @see doSend
     */
    get(apiPath, parameters, options) {
      return result.doSend(apiPath, parameters, 'GET', options);
    },

    /**
     * Send a POST request
     * @see doSend
     */
    post(apiPath, parameters, options) {
      return result.doSend(apiPath, parameters, 'POST', options);
    },

    /**
     * Send a DELETE request
     * @see doSend
     */
    delete(apiPath, parameters, options) {
      return result.doSend(apiPath, parameters, 'GET', options);
    },

    /**
     * Send a request to the server using a URL received from the API directly, receive a response
     *
     * @param {String}   $prebuiltURL       Request URL given by a previous API call
     */
    doPrebuiltSend(prebuiltURL) {
      const { headers, port } = result.prepRequest($options);

      if ($options.requester_fn) {
        const requesterOptions = {
          headers,
          method: 'GET',
          url: prebuiltURL
        };


        return $options.requester_fn(requesterOptions);
      }

      const { hostname, path } = url.parse(prebuiltURL);
      const httpsOptions = {
        headers,
        hostname,
        method: 'GET',
        path,
        port
      };

      return result.sendHttpsRequest(httpsOptions);
    },

    /**
     * Send a request to the server, receive a response
     *
     * @param {String}   apiPath       Request API path
     * @param {Object}    parameters    Parameters
     * @param {String}   _httpMethod    HTTP method to use
     * @param  {Object}    options        reconfigure the request for this call only
     */
    doSend(apiPath, parameters, _httpMethod = 'GET', options = $options) {
      const method = _httpMethod.toUpperCase();
      const { headers, hostname, port } = result.prepRequest(options);

      let query;
      const path = options.path + '/' + apiPath.replace(/\/*$/, ''); // eslint-disable-line prefer-template
      if (method === 'POST') {
        query = JSON.stringify(parameters);
        headers['Content-Type'] = 'application/json';
        headers['Content-Length'] = query.length;
      }
      else {
        query = querystring.stringify(parameters);
      }

      if (options.requester_fn) {
        const requesterOptions = {
          headers,
          hostname,
          method,
          path,
          query,
          url: `https://${hostname}${path}?${query}`
        };

        if (method === 'POST') {
          requesterOptions.body = parameters;
        }

        return options.requester_fn(requesterOptions);
      }

      const httpsOptions = {
        headers,
        hostname,
        method,
        path: `${path}?${query}`,
        port
      };

      return result.sendHttpsRequest(httpsOptions, query);
    },

    /**
     * Get a JSON response and transform to JSON
     */
    decodeResponse(response) {
      if ($options.format === 'json') {
        if (!response) {
          return {};
        }
        return JSON.parse(response);
      }

      return response;
    },

    prepRequest(options) {
      const {
        hostname: _hostname,
        http_port: httpPort,
        oauth_access_token: oauthAccessToken,
        proxy_host: proxyHost,
        proxy_port: proxyPort,
        requester_fn
      } = options;
      const hostname = !requester_fn && proxyHost ? proxyHost : _hostname;
      const port = !requester_fn && proxyHost ? proxyPort || 3128 : httpPort || 443;

      const headers = {
        'Content-Type': 'application/x-www-form-urlencoded',
        Authorization: `Bearer ${oauthAccessToken}`,
        Host: 'api.bitbucket.org',
        'User-Agent': 'NodeJS HTTP Client',
        'Content-Length': '0'
      };

      return { headers, hostname, port };
    },

    sendHttpsRequest(httpsOptions, query) {
      let resolve;
      let reject;
      const resultPromise = new Promise((_resolve, _reject) => {
        resolve = _resolve;
        reject = _reject;
      });

      const request = https.request(httpsOptions, (response) => {
        response.setEncoding('utf8');

        const rawBody = [];
        response.addListener('data', (chunk) => {
          rawBody.push(chunk);
        });
        response.addListener('end', () => {
          let body = rawBody.join('');

          if (response.statusCode >= 400) {
            if (response.headers['content-type'].includes('application/json')) {
              body = JSON.parse(body);
            }
            reject({ statusCode: response.statusCode, body });
            return;
          }

          body = result.decodeResponse(body);

          resolve({ statusCode: response.statusCode, body });
        });

        response.addListener('error', (e) => {
          reject(e);
        });

        response.addListener('timeout', () => {
          reject(new Error('Request timed out'));
        });
      });

      request.on('error', (e) => {
        reject(e);
      });

      if (httpsOptions.method === 'POST') {
        request.write(query);
      }

      request.end();

      return resultPromise;
    }
  });
};
